app/mapping.py

```diff
@@ -16,30 +16,44 @@ class MappingNotFound(BadRequest):
 
 
 class Mapping(object):
+    filter_field_prefix = "dmfilter"
+    text_search_field_prefix = "dmtext"
+    aggregatable_field_prefix = "dmagg"
+
+    # for now, these definitions are identical
+    response_field_prefix = "dmtext"
+
     def __init__(self, mapping_definition, mapping_type):
         self.definition = mapping_definition
-        self._filter_fields = tuple(sorted(
-            field.replace('filter_', '')
-            for field in self.definition['mappings'][mapping_type]['properties'].keys()
-            if field.startswith('filter_'))
-        )
-        self.filter_fields_set = frozenset(self._filter_fields)
-        self.text_fields = tuple(sorted(
-            field
-            for field in self.definition['mappings'][mapping_type]['properties'].keys()
-            if not field.startswith('filter_') and field != SERVICE_ID_HASH_FIELD_NAME
-        ))
-        self.text_fields_set = frozenset(self.text_fields)
-        self.aggregatable_fields = tuple(sorted(
-            k
-            for k, v in self.definition['mappings'][mapping_type]['properties'].items()
-            if v.get('fields', {}).get('raw', False)
-        ))
+        properties = self.definition['mappings'][mapping_type]['properties']
+
+        # build a dict of {prefix: frozenset(unprefixed_field_names)}
+        self.fields_by_prefix = {
+            prefix: frozenset(name for _, name in pairs)
+            for prefix, pairs in groupby(
+                (
+                    (prefix, maybe_name_seq[0])
+                    for prefix, *maybe_name_seq in (
+                        full_field_name.split("_", 1)
+                        for full_field_name in sorted(properties.keys())
+                    )
+                    if maybe_name_seq  # maybe_name_seq would be an empty seq if no underscores were found, discard them
+                ),
+                key=lambda x: x[0],  # (the prefix)
+            )
+        }
```

Comment:
Four comprehensions in one statement makes for a not-very-readable thing.

Reply:
But that means writing loopy, state mutating code, which is like reading assembly.

Reply:
I tend to agree that the procedural thing is just sometimes more readable. It's definitely a false equivalence / straw man argument to say that something would read like assembly just cos it's written more procedurally. Clearly opinions vary on this one, but it doesn't seem like a big ask to work on the readability here a little more, seeing as this particular function has been highlighted a couple of times now.

Reply:
I think it's a very strong analogy of procedural code to assembly - the fundamental thing in common is you mentally have to track state mutations between lines, which is a mental load you can do away with in functional declarations. Anyway, I've pulled the middle clause out into a separate, named, function now which should make it a bit clearer.